File: app/src/main/java/pl/adriankremski/coolector/repository/AuthenticationRepositoryImpl.java
package pl.adriankremski.coolector.repository;

import android.content.Context;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.functions.Function;
import pl.adriankremski.coolector.Constants;
import pl.adriankremski.coolector.TheApp;
import pl.adriankremski.coolector.model.AuthRequest;
import pl.adriankremski.coolector.model.SignUpRequest;
import pl.adriankremski.coolector.network.Api;

public class AuthenticationRepositoryImpl implements AuthenticationRepository{

    @Inject
    Api api;

    public AuthenticationRepositoryImpl(Context context) {
        TheApp.get(context).getAppComponent().inject(this);
    }

    @Override
    public Observable<String> loginWithEmail(String email, String password) {
        AuthRequest authRequest = new AuthRequest(email, password, Constants.AuthProvider.COOLECTOR);
        return api.login(authRequest).flatMap(authResponse -> Observable.just(authResponse.getToken()));
    }

    @Override
    public Observable<String> signUp(String username, String email, String password) {
        return api.signUp(new SignUpRequest(username, email, password)).flatMap(new Function<Void, ObservableSource<String>>() {
            @Override
            public ObservableSource<String> apply(Void aVoid) throws Exception {
                AuthRequest authRequest = new AuthRequest(email, password, Constants.AuthProvider.COOLECTOR);
                return api.login(authRequest).flatMap(loginResponse -> Observable.just(loginResponse.getToken()));
            }
        });
    }
}

File: app/src/main/java/pl/adriankremski/coolector/repository/AuthenticationRepositoryImpl.kt
package pl.adriankremski.coolector.repository

import android.content.Context
import io.reactivex.Observable
import pl.adriankremski.coolector.Constants
import pl.adriankremski.coolector.TheApp
import pl.adriankremski.coolector.model.AuthRequest
import pl.adriankremski.coolector.model.ResetPasswordRequest
import pl.adriankremski.coolector.model.SignUpRequest
import pl.adriankremski.coolector.network.Api
import javax.inject.Inject

class AuthenticationRepositoryImpl(context: Context) : AuthenticationRepository {
    @Inject
    lateinit var mApi: Api

    @Inject
    lateinit var mOperationRepository: OperationRepository

    init {
        TheApp[context].appComponent?.inject(this)
    }

    override fun loginWithEmail(email: String, password: String): Observable<String> {
        val authRequest = AuthRequest(email, password, Constants.AuthProvider.COOLECTOR)
        return mApi.login(authRequest).flatMap { authResponse -> Observable.just(authResponse.token) }
    }

    override fun signUp(username: String, email: String, password: String): Observable<String> {
        return mOperationRepository.pollOperation(mApi.signUp(SignUpRequest(username, email, password)))
                .flatMap {
                    val authRequest = AuthRequest(email, password, Constants.AuthProvider.COOLECTOR)
                    mApi.login(authRequest).flatMap { authResponse -> Observable.just(authResponse.token) }
                }
    }

    override fun resetPassword(email: String): Observable<Boolean> {
        return mOperationRepository.pollOperation(mApi.resetPassword(ResetPasswordRequest(email))).flatMap { Observable.just(true) }
    }
}

